File: tools/math_dsl/inverse_solver.py
"""W244 wave 32 — Newton-Raphson inverse solver for W244 kernels.

Designer workflow:
  Input:  target_rtp + kernel + fixed parameters
  Output: value of free parameter that achieves target

Replaces hand-tuning of reel weights / probabilities / pay tables with
analytic Newton-Raphson iteration. Sub-millisecond convergence in 3-8
iterations for well-conditioned kernels.

Composes with `tools.math_dsl.symbolic_gradient` for the derivative.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass(frozen=True)
class SolveResult:
    """One Newton-Raphson run outcome."""
    converged: bool
    iterations: int
    final_param: float
    final_rtp: float
    error: float
    target_rtp: float
    history: tuple[tuple[float, float], ...]  # (param, rtp) per iter


def newton_raphson_1d(
    rtp_func: Callable[[float], float],
    gradient_func: Callable[[float], float],
    target_rtp: float,
    initial_guess: float,
    *,
    tolerance: float = 1e-4,
    max_iterations: int = 30,
    param_lo: float = 0.0,
    param_hi: float = 1.0,
) -> SolveResult:
    """1-D Newton-Raphson: find param such that rtp_func(param) = target.

    Iteration: param ← param - (rtp - target) / gradient
    Bounded clamping: param stays in [param_lo, param_hi].

    Args:
      rtp_func:       maps param → rtp
      gradient_func:  maps param → ∂rtp/∂param (analytic or numerical)
      target_rtp:     RTP value to hit
      initial_guess:  starting parameter value
      tolerance:      |rtp - target| ≤ tol → converged
      max_iterations: hard cap
      param_lo/hi:    bracket for parameter
    """
    param = initial_guess
    history: list[tuple[float, float]] = []
    for i in range(max_iterations):
        rtp = rtp_func(param)
        history.append((param, rtp))
        error = abs(rtp - target_rtp)
        if error < tolerance:
            return SolveResult(
                converged=True,
                iterations=i + 1,
                final_param=param,
                final_rtp=rtp,
                error=error,
                target_rtp=target_rtp,
                history=tuple(history),
            )
        grad = gradient_func(param)
        if abs(grad) < 1e-15:
            # Stuck — gradient too small, can't progress
            break
        # Newton step
        delta = (rtp - target_rtp) / grad
        new_param = param - delta
        # Clamp to bracket
        new_param = max(param_lo, min(param_hi, new_param))
        if new_param == param:
            # Hit bracket boundary without convergence
            break
        param = new_param

    # Didn't converge within max_iterations
    rtp = rtp_func(param)
    return SolveResult(
        converged=False,
        iterations=len(history),
        final_param=param,
        final_rtp=rtp,
        error=abs(rtp - target_rtp),
        target_rtp=target_rtp,
        history=tuple(history),
    )


def bisection_1d(
    rtp_func: Callable[[float], float],
    target_rtp: float,
    *,
    param_lo: float = 0.0,
    param_hi: float = 1.0,
    tolerance: float = 1e-4,
    max_iterations: int = 50,
) -> SolveResult:
    """Robust bisection fallback (no gradient needed).

    Assumes rtp_func is monotonically increasing in param OR detects
    direction from endpoints. Useful when gradient is unreliable or
    kernel is non-smooth at boundaries.
    """
    lo = param_lo
    hi = param_hi
    rtp_lo = rtp_func(lo)
    rtp_hi = rtp_func(hi)
    history: list[tuple[float, float]] = [(lo, rtp_lo), (hi, rtp_hi)]

    # Detect direction
    increasing = rtp_hi >= rtp_lo

    # Bracket check
    if increasing:
        if not (rtp_lo <= target_rtp <= rtp_hi):
            return SolveResult(
                converged=False, iterations=2,
                final_param=lo if target_rtp < rtp_lo else hi,
                final_rtp=rtp_lo if target_rtp < rtp_lo else rtp_hi,
                error=min(abs(target_rtp - rtp_lo), abs(target_rtp - rtp_hi)),
                target_rtp=target_rtp,
                history=tuple(history),
            )
    else:
        if not (rtp_hi <= target_rtp <= rtp_lo):
            return SolveResult(
                converged=False, iterations=2,
                final_param=lo if target_rtp > rtp_lo else hi,
                final_rtp=rtp_lo if target_rtp > rtp_lo else rtp_hi,
                error=min(abs(target_rtp - rtp_lo), abs(target_rtp - rtp_hi)),
                target_rtp=target_rtp,
                history=tuple(history),
            )

    for i in range(max_iterations):
        mid = (lo + hi) / 2
        rtp_mid = rtp_func(mid)
        history.append((mid, rtp_mid))
        error = abs(rtp_mid - target_rtp)
        if error < tolerance:
            return SolveResult(
                converged=True,
                iterations=len(history),
                final_param=mid,
                final_rtp=rtp_mid,
                error=error,
                target_rtp=target_rtp,
                history=tuple(history),
            )
        if (increasing and rtp_mid < target_rtp) or (not increasing and rtp_mid > target_rtp):
            lo, rtp_lo = mid, rtp_mid
        else:
            hi, rtp_hi = mid, rtp_mid

    # Didn't converge
    mid = (lo + hi) / 2
    return SolveResult(
        converged=False,
        iterations=len(history),
        final_param=mid,
        final_rtp=rtp_func(mid),
        error=abs(rtp_func(mid) - target_rtp),
        target_rtp=target_rtp,
        history=tuple(history),
    )

File: rust-sim/src/kernels/inverse_solver.rs
//! W244.32 — inverse_solver Rust port.
//!
//! Mirror of `tools/math_dsl/inverse_solver.py`. Two convergence
//! primitives:
//!   * `newton_raphson_1d` — quadratic convergence with analytic ∂rtp/∂param.
//!   * `bisection_1d`      — robust fallback (no gradient required).
//!
//! Designer workflow:
//!   Input:  target_rtp + kernel callable + fixed parameters
//!   Output: param value that achieves target (sub-millisecond, 3-8 iter)
//!
//! Closes W244 Rust port at 19/20 kernel implementations (only
//! `showcase_game` end-to-end driver remains).

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct SolveResult {
    pub converged: bool,
    pub iterations: usize,
    pub final_param: f64,
    pub final_rtp: f64,
    pub error: f64,
    pub target_rtp: f64,
    pub history: Vec<(f64, f64)>,
}

/// 1-D Newton-Raphson: find `param` such that `rtp_func(param) == target_rtp`.
///
/// Iteration: `param ← param − (rtp − target) / gradient`, clamped to
/// `[param_lo, param_hi]`.
pub fn newton_raphson_1d<F, G>(
    rtp_func: F,
    gradient_func: G,
    target_rtp: f64,
    initial_guess: f64,
    tolerance: f64,
    max_iterations: usize,
    param_lo: f64,
    param_hi: f64,
) -> SolveResult
where
    F: Fn(f64) -> f64,
    G: Fn(f64) -> f64,
{
    let mut param = initial_guess;
    let mut history: Vec<(f64, f64)> = Vec::with_capacity(max_iterations);
    for i in 0..max_iterations {
        let rtp = rtp_func(param);
        history.push((param, rtp));
        let error = (rtp - target_rtp).abs();
        if error < tolerance {
            return SolveResult {
                converged: true,
                iterations: i + 1,
                final_param: param,
                final_rtp: rtp,
                error,
                target_rtp,
                history,
            };
        }
        let grad = gradient_func(param);
        if grad.abs() < 1e-15 {
            // Stuck — gradient too small, can't progress.
            break;
        }
        let delta = (rtp - target_rtp) / grad;
        let new_param = (param - delta).max(param_lo).min(param_hi);
        if new_param == param {
            // Hit bracket boundary without convergence.
            break;
        }
        param = new_param;
    }
    let final_rtp = rtp_func(param);
    SolveResult {
        iterations: history.len(),
        final_param: param,
        final_rtp,
        error: (final_rtp - target_rtp).abs(),
        target_rtp,
        history,
        converged: false,
    }
}

/// 1-D bisection — gradient-free, robust to non-smooth kernels.
///
/// Auto-detects monotonic direction from endpoints; refuses to run if
/// target_rtp lies outside the bracket.
pub fn bisection_1d<F>(
    rtp_func: F,
    target_rtp: f64,
    param_lo: f64,
    param_hi: f64,
    tolerance: f64,
    max_iterations: usize,
) -> SolveResult
where
    F: Fn(f64) -> f64,
{
    let mut lo = param_lo;
    let mut hi = param_hi;
    let mut rtp_lo = rtp_func(lo);
    let mut rtp_hi = rtp_func(hi);
    let mut history: Vec<(f64, f64)> = vec![(lo, rtp_lo), (hi, rtp_hi)];

    let increasing = rtp_hi >= rtp_lo;

    // Bracket check
    let in_bracket = if increasing {
        rtp_lo <= target_rtp && target_rtp <= rtp_hi
    } else {
        rtp_hi <= target_rtp && target_rtp <= rtp_lo
    };
    if !in_bracket {
        let (best_param, best_rtp) = if (target_rtp - rtp_lo).abs() <= (target_rtp - rtp_hi).abs() {
            (lo, rtp_lo)
        } else {
            (hi, rtp_hi)
        };
        return SolveResult {
            converged: false,
            iterations: 2,
            final_param: best_param,
            final_rtp: best_rtp,
            error: (best_rtp - target_rtp).abs(),
            target_rtp,
            history,
        };
    }

    for _ in 0..max_iterations {
        let mid = (lo + hi) / 2.0;
        let rtp_mid = rtp_func(mid);
        history.push((mid, rtp_mid));
        let error = (rtp_mid - target_rtp).abs();
        if error < tolerance {
            return SolveResult {
                converged: true,
                iterations: history.len(),
                final_param: mid,
                final_rtp: rtp_mid,
                error,
                target_rtp,
                history,
            };
        }
        let go_higher = (increasing && rtp_mid < target_rtp)
            || (!increasing && rtp_mid > target_rtp);
        if go_higher {
            lo = mid;
            rtp_lo = rtp_mid;
        } else {
            hi = mid;
            rtp_hi = rtp_mid;
        }
        let _ = (rtp_lo, rtp_hi); // silence unused-warning, kept for clarity
    }

    let mid = (lo + hi) / 2.0;
    let final_rtp = rtp_func(mid);
    SolveResult {
        converged: false,
        iterations: history.len(),
        final_param: mid,
        final_rtp,
        error: (final_rtp - target_rtp).abs(),
        target_rtp,
        history,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Charge-meter shape: rtp = p × 10 with target 0.10 → p = 0.5.
    fn charge_meter_rtp(p: f64) -> f64 {
        p * 0.20 // expected_charge_per_spin × award/threshold = 0.5×10/50
    }

    fn charge_meter_grad(_p: f64) -> f64 {
        0.20
    }

    #[test]
    fn newton_finds_charge_meter_target_in_one_iter() {
        let res = newton_raphson_1d(
            charge_meter_rtp,
            charge_meter_grad,
            0.10,    // target
            0.3,     // initial guess
            1e-4,
            30,
            0.0, 1.0,
        );
        assert!(res.converged);
        assert!((res.final_param - 0.5).abs() < 1e-4);
        // Quadratic convergence: should hit target on 2nd or 3rd iter.
        assert!(res.iterations <= 3);
    }

    #[test]
    fn newton_clamps_to_bracket() {
        // Force overshoot: tiny gradient on very off-target.
        let res = newton_raphson_1d(
            |p| p,
            |_p| 0.01, // tiny gradient → big delta
            0.5,
            0.99,
            1e-6,
            5,
            0.0,
            1.0,
        );
        // Hit boundary, didn't converge (param stuck at clamp).
        assert!(res.final_param >= 0.0 && res.final_param <= 1.0);
    }

    #[test]
    fn bisection_finds_monotone_increasing_target() {
        // pay_anywhere shape: rtp ∝ p (increasing)
        let res = bisection_1d(
            |p| p * 0.5,
            0.20,
            0.0, 1.0,
            1e-4,
            50,
        );
        assert!(res.converged);
        assert!((res.final_param - 0.4).abs() < 1e-3);
    }

    #[test]
    fn bisection_handles_decreasing_direction() {
        let res = bisection_1d(
            |p| 1.0 - p, // decreasing in p
            0.4,
            0.0, 1.0,
            1e-4,
            50,
        );
        assert!(res.converged);
        assert!((res.final_param - 0.6).abs() < 1e-3);
    }

    #[test]
    fn bisection_refuses_out_of_bracket_target() {
        let res = bisection_1d(
            |p| p,
            2.0, // target outside [0, 1] image
            0.0, 1.0,
            1e-4,
            50,
        );
        assert!(!res.converged);
    }
}
